### ISSUE TITLE
Slate data model: decouple structure and UI aspects.

### ISSUE THREAD

**Issue by bebbi** [ianstormtaylor/slate#2552]
#### Do you want to request a _feature_ or report a _bug_?

Discussion

I'd like to raise an opportunity for clarifying the docs based on a slack chat.

The docs state that a custom `renderNode` function needs to mix in the slate dom `attributes`.

Now, the slate data model has a dual role: structural (navigating/editing/normalizing), and UI (dom rendering).

I've been looking for ways to decouple these roles.
For example, in the tree-based model I need to build, I find a substantial code simplification by always wrapping certain children nodes in a `type: children` container. However, due to html restrictions (such as `td`'s required to be inside `tr` directly without any intermediate wrappers) I cannot always render a direct dom equivalent for these `children` nodes.

In summary, while it's possible to model out all render details in the slate model, in practice, access to the slate data model can much simpler if we could treat some nodes as "model only" and not require all to render.

I have had success for my use-case that needs situational rendering of such `model` nodes by simply skipping node rendering:

    renderNode(props, ...args) {
      if (someCondition() {
        // no props.attributes mixed in!
        return props.children
      }
    }

But I wonder if that's supposed to work? Is there any guidance on if/when nodes can skip dom representation, and when they cannot?
I'll happily add a PR to the doc if there's any expert wisdom coming out of this.

**Comment by ianstormtaylor** [comment]
@bebbi what's your use case here? I can't think of why you'd want this.

**Comment by bebbi** [comment]
@ianstormtaylor my simplest use-case is: 'track change' functionality like in similar systems, where you can show or hide proposed inline deletions, for example with a strikethrough on showing.

I'd like to make this a pure UI concern and keep the `deletes` in the nodes and just render `null` when I don't show them. Like that they could easily move around on edits. Slate doesn't seem to allow this as it always wants to see slate dom attributes for all its nodes, is that correct?

**Comment by ianstormtaylor** [comment]
Not sure to be honest. Could you use `display: none` instead so that it still lived in the DOM for reconciliation but wasn’t visible? Just an idea. I haven’t done any testing with invisible content.

**Comment by davidhoeller** [comment]
I'm also interested in what @bebbi proposes. I need to have 'collapsed', i.e. hidden content (blocks and inlines). For example, I'm collapsing several nodes in one "minimized" representation (which would be one node), so the content consists of rendered and not-rendered nodes. Th e not-rendered nodes still should be selectable, copyable, pasteable, similar to @bebbi's description – also (and especially) if such collapsed nodes are part of "mixed" expanded/collapsed content (the rended nodes surrounding the non-rendered being selected). 

At the moment, I'm removing the collapsed content, hiding it from Slate by deleting the nodes, and storing their JSON representation in the "minimized" node's data – far from ideal. I'm trying to prevent the "Slat can't find DOM node" errors and end up losing Slate's control over the content. Haven't yet had a better idea, though.

So, optional rendering of nodes and maybe a more flexible mapping between model and DOM would be a great direction to think and talk …

@ianstormtaylor I've also tried the display:none option but ended up with incredibly polluted DOM and also I could not select past the hidden content. Also I have tha feeling that does not scale if we have lots of hidden content …

**Comment by bebbi** [comment]
I tried this:

1. returning `null` or `display:none`: no dom -> `Unable to find a DOM node for`.. error
2.

```
return (
  <span
    {...attributes}
    contentEditable={false}
    style={{ visibility: 'hidden', height: 0, width: 0, overflow: 'hidden' }}
  >
    {children}
  </span>)
```
or a variant with a non-editable outer span; both hide the dom but still manage the cursor as if it was present.

**Comment by ianstormtaylor** [comment]
I think this is something that we're not going to change. And instead people should work around it with React-specific ways of rendering children/wrapper components/etc.

**Comment by bebbi** [comment]
@ianstormtaylor thanks for clarifying - one prominent use-case that seems to fall into this category though would certainly benefit from at least your advice:

How about a node that nests others when I want to display that node in 'collapsed' mode, i.e. not showing its descendants?

As far as I understand, the current approach allows for displaying children using React in arbitrary ways, but it doesn't allow not rendering them - because selecting a range would complain about missing DOM attributes.
Adding and removing descendants from the slate model, depending on visibility while not breaking undo etc. sounds like a hard thing to do?

**Comment by ianstormtaylor** [comment]
You might be able to approximate it by using a dynamic `isVoid` query and make collapsed nodes voids.
